configure.ac

```
@@ -6686,11 +6686,10 @@ int main(void)
 ]])],
 [ac_cv_strftime_c99_support=yes],
 [ac_cv_strftime_c99_support=no],
-[ac_cv_strftime_c99_support=no])])
-if test "$ac_cv_strftime_c99_support" = yes
+[ac_cv_strftime_c99_support=yes])])
+if test "$ac_cv_strftime_c99_support" = no
 then
-  AC_DEFINE([Py_STRFTIME_C99_SUPPORT], [1],
-  [Define if C99-specific strftime specifiers are supported.])
+  AC_MSG_ERROR([Python requires C99-compatible strftime specifiers])
 fi
```

Let's just simplify this to adding the `AC_MSG_ERROR` in the _action-if-false_ clause of `AC_RUN_IFELSE`. (I think I got the suggestion right; please double check)

```suggestion
[],
[AC_MSG_ERROR([Python requires C99-compatible strftime specifiers])])])
```

---

Great! I think I tried that, but don't recall what went wrong. I'll add the patch and test it.

---

Ah okay; this patch works fine, but how does it interact with the `AC_CACHE_CHECK([whether C99-compatible strftime specifiers are supported], [ac_cv_strftime_c99_support], [ ...` bit above?

---

We can drop that too, though I don't know if the caching matters?

```diff
diff --git a/configure.ac b/configure.ac
index ccb0549dc33..cad36ba1011 100644
--- a/configure.ac
+++ b/configure.ac
@@ -6665,7 +6665,6 @@ then
   [Define if year with century should be normalized for strftime.])
 fi
 
-AC_CACHE_CHECK([whether C99-compatible strftime specifiers are supported], [ac_cv_strftime_c99_support], [
 AC_RUN_IFELSE([AC_LANG_SOURCE([[
 #include <time.h>
 #include <string.h>
@@ -6685,7 +6684,7 @@ int main(void)
 }
 ]])],
 [],
-[AC_MSG_ERROR([Python requires C99-compatible strftime specifiers])])])
+[AC_MSG_ERROR([Python requires C99-compatible strftime specifiers])])
 
 dnl check for ncursesw/ncurses and panelw/panel
 dnl NOTE: old curses is not detected.
```

---

(I pushed these changes)

---

So.. removing the caching does have an effect, it looks like it broke the WASI CI build

Caching can be used for cross-compiles discussed in https://www.gnu.org/software/autoconf/manual/autoconf-2.63/html_node/Runtime.html

> To configure for cross-compiling you can also choose a value for those parameters based on the canonical system name (see [Manual Configuration](https://www.gnu.org/software/autoconf/manual/autoconf-2.63/html_node/Manual-Configuration.html#Manual-Configuration)). Alternatively, set up a test results cache file with the correct values for the host system (see [Caching Results](https://www.gnu.org/software/autoconf/manual/autoconf-2.63/html_node/Caching-Results.html#Caching-Results)).

I'll see if adding an empty `action-if-cross-compiling` block changes the behavior

---

https://github.com/python/cpython/pull/128106/commits/7928672470c7d606d5554d22f3b3a60603966d95 doesn't change the `configure` file, so it shouldn't have any affect. 

I think we'll need to skip this explicitly when cross-compiling

---

The skip would look like this
```diff
diff --git a/configure b/configure
index 31af8b060c0..4ee003e0b21 100755
--- a/configure
+++ b/configure
@@ -26395,6 +26395,9 @@ printf "%s\n" "#define Py_NORMALIZE_CENTURY 1" >>confdefs.h
 
 fi
 
+if test "x$cross_compiling" = xyes
+then :
+
 if test "$cross_compiling" = yes
 then :
   { { printf "%s\n" "$as_me:${as_lineno-$LINENO}: error: in \`$ac_pwd':" >&5
@@ -26433,6 +26436,7 @@ rm -f core *.core core.conftest.* gmon.out bb.out conftest$ac_exeext \
   conftest.$ac_objext conftest.beam conftest.$ac_ext
 fi
 
+fi
 
 have_curses=no
 have_panel=no
diff --git a/configure.ac b/configure.ac
index cad36ba1011..0160e9d4c7e 100644
--- a/configure.ac
+++ b/configure.ac
@@ -6665,6 +6665,7 @@ then
   [Define if year with century should be normalized for strftime.])
 fi
 
+AS_VAR_IF([cross_compiling], [yes], [
 AC_RUN_IFELSE([AC_LANG_SOURCE([[
 #include <time.h>
 #include <string.h>
@@ -6684,7 +6685,7 @@ int main(void)
 }
 ]])],
 [],
-[AC_MSG_ERROR([Python requires C99-compatible strftime specifiers])])
+[AC_MSG_ERROR([Python requires C99-compatible strftime specifiers])])])
 
 dnl check for ncursesw/ncurses and panelw/panel
 dnl NOTE: old curses is not detected.
 ```
 
 Curious for your thoughts. Retaining the cache seems most consistent with other uses of `AC_RUN_IFELSE` but I'm not sure if it's intentionally setup for cross-compiling purposes.

---

Please leave the cache check as it was; there should be no reason to re-check this when we're running `configure -C`.

---

Then we need to set the variable still, right?

---

i.e., we should not apply the simplifying suggestion here?

---

> Then we need to set the variable still, right?

Yes; setting the _action-if-cross-compile_ to the empty string `[]` should be sufficient.

---

(You need to regen configure.)

---

Okay I think we're on the same page now with one minor detail, setting `[]` for cross-compiles doesn't change the `configure` file so I don't think it has any effect:

```
❯ gd
diff --git a/configure.ac b/configure.ac
index 957ce237bb9..8c8a403bd99 100644
--- a/configure.ac
+++ b/configure.ac
@@ -6685,8 +6685,7 @@ int main(void)
 }
 ]])],
 [ac_cv_strftime_c99_support=yes],
-[AC_MSG_ERROR([Python requires C99-compatible strftime specifiers])],
-[])])
+[AC_MSG_ERROR([Python requires C99-compatible strftime specifiers])])])
 
 dnl check for ncursesw/ncurses and panelw/panel
 dnl NOTE: old curses is not detected.
❯ make regen-configure
./Tools/build/regen-configure.sh
+ IMAGE=ghcr.io/python/autoconf:2024.11.11.11786316759
+ AUTORECONF='autoreconf -ivf -Werror'
+ WORK_DIR=/src
+++ dirname ./Tools/build/regen-configure.sh
++ cd ./Tools/build/../..
++ pwd
+ abs_srcdir=/Users/zb/workspace/cpython
+ podman --version
+ RUNTIME=podman
+ PATH_OPT=
+ command -v selinuxenabled
+ podman run --rm -v /Users/zb/workspace/cpython:/src ghcr.io/python/autoconf:2024.11.11.11786316759
Rebuilding configure script using autoconf (GNU Autoconf) 2.71
autoreconf: export WARNINGS=error
autoreconf: Entering directory '.'
autoreconf: configure.ac: not using Gettext
autoreconf: running: aclocal --force 
autoreconf: configure.ac: tracing
autoreconf: configure.ac: not using Libtool
autoreconf: configure.ac: not using Intltool
autoreconf: configure.ac: not using Gtkdoc
autoreconf: running: /usr/local/bin/autoconf --force
autoreconf: running: /usr/local/bin/autoheader --force
autoreconf: configure.ac: not using Automake
autoreconf: Leaving directory '.'
❯ gd
diff --git a/configure.ac b/configure.ac
index 957ce237bb9..8c8a403bd99 100644
--- a/configure.ac
+++ b/configure.ac
@@ -6685,8 +6685,7 @@ int main(void)
 }
 ]])],
 [ac_cv_strftime_c99_support=yes],
-[AC_MSG_ERROR([Python requires C99-compatible strftime specifiers])],
-[])])
+[AC_MSG_ERROR([Python requires C99-compatible strftime specifiers])])])
 
 dnl check for ncursesw/ncurses and panelw/panel
 dnl NOTE: old curses is not detected.
 ```

---

However, the WASI job is still broken: https://github.com/python/cpython/actions/runs/12588980254/job/35087957232?pr=128106

---

Aaaargh, autoconf #!"$% :) Ok, let's just apply this and land this PR:

```diff
diff --git a/configure b/configure
index eec490902f9..7108f340be9 100755
--- a/configure
+++ b/configure
@@ -26404,10 +26404,7 @@ else $as_nop
 
 if test "$cross_compiling" = yes
 then :
-  { { printf "%s\n" "$as_me:${as_lineno-$LINENO}: error: in \`$ac_pwd':" >&5
-printf "%s\n" "$as_me: error: in \`$ac_pwd':" >&2;}
-as_fn_error $? "cannot run test program while cross compiling
-See \`config.log' for more details" "$LINENO" 5; }
+  ac_cv_strftime_c99_support=yes
 else $as_nop
   cat confdefs.h - <<_ACEOF >conftest.$ac_ext
 /* end confdefs.h.  */
diff --git a/configure.ac b/configure.ac
index 957ce237bb9..41b082e6d3b 100644
--- a/configure.ac
+++ b/configure.ac
@@ -6686,7 +6686,7 @@ int main(void)
 ]])],
 [ac_cv_strftime_c99_support=yes],
 [AC_MSG_ERROR([Python requires C99-compatible strftime specifiers])],
-[])])
+[ac_cv_strftime_c99_support=yes])])
 
 dnl check for ncursesw/ncurses and panelw/panel
 dnl NOTE: old curses is not detected.
```

---

I'll see if we can get away with an empty value https://github.com/python/cpython/pull/128106/commits/67ce5393fe1b240bb1d1c1e06daa651db9af4b8a — which I believe is the correct convention if we can't determine the value.

If that doesn't work, I'll just use `yes` and can make a note to myself to revisit it once I figure out how to reproduce the WASI build locally :)

---

Victory!